internal/server/v2/observation/calculation.go

```diff
@@ -17,56 +17,137 @@ package observation
+		facetObs := []*pbv2.FacetObservation{}
+		for facetId, facetData := range entityData {
+			if _, ok := newResp.Facets[facetId]; !ok {
+				newResp.Facets[facetId] = variableResp.Facets[facetId]
+			}
+			facetObs = append(facetObs, &pbv2.FacetObservation{
+				FacetId:      facetId,
+				Observations: facetData,
+				EarliestDate: facetData[0].GetDate(),
+				LatestDate:   facetData[len(facetData)-1].GetDate(),
+				ObsCount:     int32(len(facetData)),
+			})
+		}
+		if len(newResp.ByVariable[variable].ByEntity) == 0 {
+			newResp.ByVariable[variable].ByEntity = map[string]*pbv2.EntityObservation{}
+		}
+		newResp.ByVariable[variable].ByEntity[entity] = &pbv2.EntityObservation{
+			OrderedFacets: facetObs,
+		}
+	}
+	return newResp, nil
+}
+
+// Detects holes in a V2ObservationResponse and attempts to fill them using calculations.
+func MaybeCalculateHoles(
+	ctx context.Context,
+	store *store.Store,
+	cachedata *cache.Cache,
+	metadata *resource.Metadata,
+	httpClient *http.Client,
+	inputReq *pbv2.ObservationRequest,
+	inputResp *pbv2.ObservationResponse,
+) ([]*pbv2.ObservationResponse, error) {
+	result := []*pbv2.ObservationResponse{}
+	holes, err := findObservationResponseHoles(inputReq, inputResp)
+	if err != nil {
+		return nil, err
+	}
+	for variable, entity := range holes {
 		formulas, ok := cachedata.SVFormula()[variable]
 		if !ok {
 			continue
 		}
-		for entity, entityObservation := range variableObservation.ByEntity {
-			// Response already contains data.
-			if len(entityObservation.OrderedFacets) != 0 {
-				continue
+		currentEntity := entity
+		for _, formula := range formulas {
+			calculatedResp, err := Calculate(
+				ctx,
+				store,
+				cachedata,
+				metadata,
+				httpClient,
+				formula,
+				variable,
+				currentEntity,
+				inputReq,
+			)
+			if err != nil {
+				return nil, err
 			}
-			// Use first formula that returns data.
-			for _, formula := range formulas {
-				derivedSeries, err := DerivedSeries(
-					ctx,
-					store,
-					formula,
-					[]string{entity},
-				)
-				// Missing input data.
-				if err != nil {
-					continue
+			variableObs := calculatedResp.ByVariable[variable]
+			if entity.GetExpression() != "" {
+				if len(variableObs.ByEntity) > 0 {
+					result = append(result, calculatedResp)
+					break
+				}
+			} else if len(entity.GetDcids()) != 0 {
+				emptyDcids := []string{}
+				for _, dcid := range entity.GetDcids() {
+					if _, ok := variableObs.ByEntity[dcid]; !ok {
+						emptyDcids = append(emptyDcids, dcid)
+					}
 				}
-				// Successful calculation.
-				bv := derivedSeries.ByVariable
-				if len(bv[formula].ByEntity[entity].OrderedFacets) > 0 {
-					// Re-label formula response with the outputProperty and delete
-					// formula.
-					bv[variable] = bv[formula]
-					delete(bv, formula)
-					calculatedResponses = append(calculatedResponses, derivedSeries)
-					continue
+				if len(emptyDcids) < len(entity.GetDcids()) {
```

Comment:
is it possible that len(emptyDcids) > entity.GetDcids()? i wonder if you can remove the check of line 141

Reply:
It's not possible to be greater but it can be equal (i.e. nothing got computed from this round), so I didn't want to keep appending empty responses to result in this case